File: frontend/src/services/brandService.ts
import { apiClient } from "./apiClient";
import { resolveMediaUrl } from "@/utils/media";
import type { Brand, BrandFormValues } from "@/types/brand";

function withResolvedLogo(brand: Brand): Brand {
  return { ...brand, logoUrl: resolveMediaUrl(brand.logoUrl) };
}

export const brandService = {
  async list(includeInactive = false): Promise<Brand[]> {
    const { data } = await apiClient.get<Brand[]>("/brands", { params: { includeInactive } });
    return data.map(withResolvedLogo);
  },

  async create(values: BrandFormValues): Promise<Brand> {
    const { data } = await apiClient.post<Brand>("/brands", values);
    return withResolvedLogo(data);
  },

  async update(id: number, values: BrandFormValues): Promise<Brand> {
    const { data } = await apiClient.put<Brand>(`/brands/${id}`, values);
    return withResolvedLogo(data);
  },

  async remove(id: number): Promise<void> {
    await apiClient.delete(`/brands/${id}`);
  },
};

File: backend/src/main/java/com/concessionaria/brand/BrandService.java
package com.concessionaria.brand;

import com.concessionaria.brand.dto.BrandRequest;
import com.concessionaria.brand.dto.BrandResponse;
import com.concessionaria.exception.BusinessRuleException;
import com.concessionaria.exception.DuplicateResourceException;
import com.concessionaria.exception.ResourceNotFoundException;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional(readOnly = true)
public class BrandService {

    private final BrandRepository brandRepository;

    public BrandService(BrandRepository brandRepository) {
        this.brandRepository = brandRepository;
    }

    public List<BrandResponse> findAll(boolean includeInactive) {
        List<Brand> brands = includeInactive
                ? brandRepository.findAll(org.springframework.data.domain.Sort.by("name"))
                : brandRepository.findByActiveTrueOrderByNameAsc();
        return brands.stream().map(BrandResponse::from).toList();
    }

    public BrandResponse findById(Long id) {
        return BrandResponse.from(getOrThrow(id));
    }

    @Transactional
    public BrandResponse create(BrandRequest request) {
        if (brandRepository.existsByNameIgnoreCase(request.name())) {
            throw new DuplicateResourceException("Já existe uma marca com este nome.");
        }
        Brand brand = new Brand();
        applyRequest(brand, request);
        return BrandResponse.from(brandRepository.save(brand));
    }

    @Transactional
    public BrandResponse update(Long id, BrandRequest request) {
        Brand brand = getOrThrow(id);
        if (brandRepository.existsByNameIgnoreCaseAndIdNot(request.name(), id)) {
            throw new DuplicateResourceException("Já existe uma marca com este nome.");
        }
        applyRequest(brand, request);
        return BrandResponse.from(brandRepository.save(brand));
    }

    @Transactional
    public void delete(Long id) {
        Brand brand = getOrThrow(id);
        try {
            brandRepository.delete(brand);
            brandRepository.flush();
        } catch (DataIntegrityViolationException e) {
            throw new BusinessRuleException("Não é possível excluir uma marca com veículos vinculados.");
        }
    }

    private void applyRequest(Brand brand, BrandRequest request) {
        brand.setName(request.name());
        brand.setLogoUrl(request.logoUrl());
        if (request.active() != null) {
            brand.setActive(request.active());
        }
    }

    private Brand getOrThrow(Long id) {
        return brandRepository.findById(id)
                .orElseThrow(() -> ResourceNotFoundException.of("Marca", id));
    }
}
